src/test/java/com/android/volley/toolbox/DiskBasedAsyncCacheTest.java

```diff
@@ -0,0 +1,449 @@
+
+        putEntry("entry4", entry4).get();
+
+        assertThat(getEntry("entry1").get(), is(nullValue()));
+        assertThat(getEntry("entry2").get(), is(nullValue()));
+        assertThatEntriesAreEqual(getEntry("entry3").get(), entry3);
+        assertThatEntriesAreEqual(getEntry("entry4").get(), entry4);
+    }
+
+    @Test
+    public void testGetBadMagic() throws IOException, ExecutionException, InterruptedException {
+        // Cache something
+        Cache.Entry entry = randomData(1023);
+        putEntry("key", entry).get();
+        assertThatEntriesAreEqual(getEntry("key").get(), entry);
+
+        // Overwrite the magic header
+        File cacheFolder = temporaryFolder.getRoot();
+        File file = cacheFolder.listFiles()[0];
+        FileOutputStream fos = new FileOutputStream(file);
+        DiskBasedCacheUtility.writeInt(fos, 0);
+        fos.close();
+        assertThat(getEntry("key").get(), is(nullValue()));
+        assertThat(listCachedFiles(), is(emptyArray())); // should this be??
+    }
+
+    @Test
+    public void testGetWrongKey() throws IOException, ExecutionException, InterruptedException {
+        // Cache something
+        Cache.Entry entry = randomData(1023);
+        putEntry("key", entry).get();
+        assertThatEntriesAreEqual(getEntry("key").get(), entry);
+
+        // Access the cached file
+        File cacheFolder = temporaryFolder.getRoot();
+        File file = cacheFolder.listFiles()[0];
+
+        // Write a new header to file associated with key
+        AsynchronousFileChannel afc =
+                AsynchronousFileChannel.open(file.toPath(), StandardOpenOption.WRITE);
+        CacheHeader wrongHeader = new CacheHeader("bad", entry);
+        ByteBuffer buffer = ByteBuffer.allocate(59);
+        wrongHeader.writeHeader(buffer);
+        buffer.flip();
+        Future<Integer> operation = afc.write(buffer, 0);
+        operation.get();
+        afc.close();
+
+        // key is gone, but file is still there
+        assertThat(getEntry("key").get(), is(nullValue()));
+        assertThat(listCachedFiles(), is(arrayWithSize(1)));
+
+        // Note: file is now a zombie because its key does not map to its name
+    }
+
+    @Test
+    public void testPutRemoveGet() throws ExecutionException, InterruptedException {
+        Cache.Entry entry = randomData(511);
+        putEntry("key", entry).get();
+
+        assertThatEntriesAreEqual(getEntry("key").get(), entry);
+
+        removeEntry("key").get();
+        assertThat(getEntry("key").get(), is(nullValue()));
+        assertThat(listCachedFiles(), is(emptyArray()));
+    }
+
+    @Test
+    public void testPutClearGet() throws ExecutionException, InterruptedException {
+        Cache.Entry entry = randomData(511);
+        putEntry("key", entry).get();
+
+        assertThatEntriesAreEqual(getEntry("key").get(), entry);
+
+        cache.clear(
+                new AsyncCache.OnCompleteCallback() {
```

Comment:
Why did you decide to do the asserts in the callback this time instead of using a future?

Reply:
I thought it'd be better since it's only only one level of indentation, and it'd be more code to use the future. But it definitely is less readable, now that I look at it again. I just changed it.